### PR TITLE
[GLIB] Don't require GApplication for bwrap sandbox

### PR DESCRIPTION
**Pull request by TingPing**
#### 2bab2a58e316087dbe1bb5aa601f5e57d4384801
<pre>
[GLIB] Don&apos;t require GApplication for bwrap sandbox
<a href="https://bugs.webkit.org/show_bug.cgi?id=255249">https://bugs.webkit.org/show_bug.cgi?id=255249</a>

Reviewed by Adrian Perez de Castro and Carlos Garcia Campos.

The app-id isn&apos;t that important to our usage in our bubblewrap sandbox
so instead of erroring we can generate an id without any real downsides.

* Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
(WebKit::applicationId):
(WebKit::createFlatpakInfo):

Canonical link: <a href="https://commits.webkit.org/262947@main">https://commits.webkit.org/262947@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/49d2edfe2132b58368cc491ecc7c9d2ed05ab0bf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2746 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2745 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2892 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4155 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3109 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2890 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2844 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2415 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2772 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3203 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2457 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3925 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/697 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2442 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2538 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2453 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2493 "10 flakes 5 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3677 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2826 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2270 "5 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2489 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2445 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2482 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/364 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2639 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp [hunk 1]
```diff
@@ -124,20 +125,24 @@ int argumentsToFileDescriptor(const Vector<CString>& args, const char* name)
     return memfd;
 }
 
-static const char* applicationId(GError** error)
+static String applicationId()
```
Comment (shivamidow) [inline Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: /* Unofficial Review */
Suggestion: ensureApplicationId()? Because this function always returns an application id.

File: Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp [hunk 2]
```diff
@@ -124,20 +125,24 @@ int argumentsToFileDescriptor(const Vector<CString>& args, const char* name)
...
 {
     GApplication* app = g_application_get_default();
-    if (!app) {
-        g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "GApplication is required.");
-        return nullptr;
+    if (app) {
+        const char* appID = g_application_get_application_id(app);
+        if (appID)
+            return String::fromUTF8(appID);
```
Comment (shivamidow) [inline Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: /* Unofficial Review */
```
if (auto* app = g_application_get_default()) {
    if (const auto* appID = g_application_get_application_id(app))
        return String::fromUTF8(appID);
}
```

File: Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp [hunk 3]
```diff
@@ -124,20 +125,24 @@ int argumentsToFileDescriptor(const Vector<CString>& args, const char* name)
...
-        return nullptr;
-    }
-    return appID;
+    const char* programName = g_get_prgname();
+    if (programName && g_application_id_is_valid(programName))
+        return String::fromUTF8(programName);
+
+    // There must be some id for xdg-desktop-portal to function.
+    // While xdg-desktop-portal does use this for permissions it's value
+    // doesn't matter much in our usage as we don't use any permissioned APIs.
```
Comment (shivamidow) [inline Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: /* Unofficial Review */
Suggestion:
We create a random ID for permission of xdg-desktop-portal, but the ID is not necessarily valid since we don't use permission-required APIs.

File: Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp [hunk 4]
```diff
@@ -124,20 +125,24 @@ int argumentsToFileDescriptor(const Vector<CString>& args, const char* name)
...
 }
 
-static const char* applicationId(GError** error)
+static String applicationId()
 {
     GApplication* app = g_application_get_default();
-    if (!app) {
-        g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "GApplication is required.");
-        return nullptr;
+    if (app) {
```
Comment (carlosgcampos) [inline Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: if (auto* app = g_application_get_default()) {

File: Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp [hunk 5]
```diff
@@ -124,20 +125,24 @@ int argumentsToFileDescriptor(const Vector<CString>& args, const char* name)
...
+        const char* appID = g_application_get_application_id(app);
+        if (appID)
```
Comment (carlosgcampos) [inline Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]:
```suggestion
        if (const char* appID = g_application_get_application_id(app))
```

**Review by carlosgcampos — APPROVED** [review]

File: Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp [hunk 6]
```diff
@@ -146,15 +151,9 @@ static int createFlatpakInfo()
...
-        // them to be application unique which is best done via GApplication.
-        GUniqueOutPtr<GError> error;
-        const char* appID = applicationId(&error.outPtr());
-        if (!appID)
-            g_error("Unable to configure xdg-desktop-portal access in the WebKit sandbox: %s", error->message);
-
+        String appID = applicationId();
         GUniquePtr<GKeyFile> keyFile(g_key_file_new());
-        g_key_file_set_string(keyFile.get(), "Application", "name", appID);
+        g_key_file_set_string(keyFile.get(), "Application", "name", appID.utf8().data());
```
Comment (shivamidow) [inline Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: /* Unofficial Review */
appID.utf8().data() -> ensureApplicationID().utf8().data(). Then, we can remove the declaration of appID above.

By the way, the only user of the applicationID() is here. char* -> String -> String (copy) ->  char* is inefficient. I suggest keeping const char* instead of String for the return value.

**Review by shivamidow — COMMENTED** [review]
Not sure of the necessity of this change. But this is an unofficial review, so I defer the judgement to formal reviewers.

**Comment by aperezdc** [comment]
I like the overall idea of this patch. One thought: Would it make sense to try and see if `g_get_prgname()` returns something usable (non null, non empty) and using it with its contents sanitized to make sure the resulting identifier is valid; then generating a random UUID would only be used as fallback if both using `GApplication` and `g_get_prgname()` aren't possible. Or am I missing some reason why this wouldn't be a good idea?

**Comment by carlosgcampos** [comment]
> I like the overall idea of this patch. One thought: Would it make sense to try and see if `g_get_prgname()` returns something usable (non null, non empty) and using it with its contents sanitized to make sure the resulting identifier is valid; then generating a random UUID would only be used as fallback if both using `GApplication` and `g_get_prgname()` aren't possible. Or am I missing some reason why this wouldn't be a good idea?

This is what the patch is doing, first tries with GAppliation, then prgname and then UUID

**Review by aperezdc — APPROVED** [review]

File: Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp [hunk 7]
```diff
@@ -124,20 +125,24 @@ int argumentsToFileDescriptor(const Vector<CString>& args, const char* name)
...
-        g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "GApplication must have a valid ID.");
-        return nullptr;
-    }
-    return appID;
+    const char* programName = g_get_prgname();
+    if (programName && g_application_id_is_valid(programName))
+        return String::fromUTF8(programName);
+
+    // There must be some id for xdg-desktop-portal to function.
+    // While xdg-desktop-portal does use this for permissions it's value
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]:
```suggestion
    // While xdg-desktop-portal does use this for permissions its value
```

File: Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp (hunk 3 above)
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: I highly doubt we can actually guarantee this comment remains true going forward, so I would reword it to instead say the best we can do is use a new app ID with new permissions rather than say that we don't do anything that requires checking permissions.

However, I worry that by using a new UUID each time we may bloat the permissions store. It's probably designed with the expectation that there will be a small number of apps. We should probably store the UUID in the data directory to avoid this. Bonus: then it will be persistent, and permissions will work properly.

File: Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp (hunk 6 above)
Comment (TingPing) [reply Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: The API is no longer `const char*` because it dynamically generates a string, so a copy will always be around. `String` vs `GUniquePtr<char>` isn't a big difference IMO.

File: Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp (hunk 3 above)
Comment (TingPing) [reply Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: I agree with the wording.

I really don't want to start storing this string though. Nothing should be added to the permission store.

The types of features that have permissions are things that a web content process should never do. Like screen sharing or file choosers. These are always going to be handled by other processes in WebKit which does its own permissions.

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: Feel fragile, because there's no way to enforce it. We won't notice if we start doing something that requires permissions, or if something we currently do that does not require a permission starts to do so in the future. The portal APIs don't have a way to assert that they'll never prompt for permission or to indicate that all permissions should be denied. But maybe we should add one....

Comment (TingPing) [reply Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: I'm going to merge this for now and open an issue to track this idea if that's OK?

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: I suppose, though I don't see what's so bad about storing the app ID. It's not exactly a big file.

Comment (aperezdc) [reply Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: Well, storing the identifiers opens a new can of worms: When are the files deleted? What if writing to disk fails (out of disk space, or so)? What if files disappear between runs? And so on. I would rather forget about saving the identifiers for now and see how things go. We can always make a follow-up patch later if needed.

Alternatively, it could be better to open `/proc/self/exe`, calculate the SHA256 checksum of the file, and use it as stable identifier: as long as the program on disk doesn't change, the identifier will be the same without needing to store anything else on disk... But I digress, even that I would leave for a follow up. If needed.

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: This might come back to bite us in a few years when our permission stores grow huge, though. Imagine if your cookies database permanently increased in size each time you launch the browser. Using a random app ID results in the same thing, except for portal permissions.

Proposed alternative: use one hardcoded ID so that all applications get the exact same fallback ID and all permissions are shared.

Comment (TingPing) [reply Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: > Proposed alternative: use one hardcoded ID so that all applications get the exact same fallback ID and all permissions are shared.

If WebKitGTK happens to use a permissioned API, which is when this theoretical problem appears, then this becomes a security risk. If you grant Webcam/ScreenSharing/etc permissions to one application you grant it to all of them.

Comment (TingPing) [reply Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: > Alternatively, it could be better to open /proc/self/exe, calculate the SHA256 checksum of the file, and use it as stable identifier: as long as the program on disk doesn't change, the identifier will be the same without needing to store anything else on disk

This sounds like a decent idea to me.

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: I suppose. A new permission store entry every time the binary changes is much less bad than a new one every time you start the program.

Comment (TingPing) [reply Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp]: https://github.com/WebKit/WebKit/pull/12761
